src/react/provider.js

```diff
@@ -1,16 +1,42 @@
-import React, { useContext, useMemo, forwardRef, useEffect } from 'react';
+import React, { forwardRef, Component } from 'react';
 import PropTypes from 'prop-types';
 
 import { diRegistry } from './constants';
 import { Context } from './context';
 import { addInjectableToMap, getDisplayName, findInjectable } from './utils';
 import { globalDi } from './global';
 
-export const DiProvider = ({ children, use, target, global }) => {
-  const { getDependencies } = useContext(Context);
+export class DiProvider extends Component {
+  static contextType = Context;
+
+  static propTypes = {
+    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
+    global: PropTypes.bool,
+    target: PropTypes.oneOfType([
+      PropTypes.func,
+      PropTypes.arrayOf(PropTypes.func),
+    ]),
+    use: PropTypes.arrayOf(
+      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
+    ).isRequired,
+  };
+
+  componentDidCatch(err) {
+    globalDi._remove(this.props.use);
+    throw err;
```

Comment:
Just to double check - :notsureif: unmount not happens during catch.
I have a very string dejavu feeling right now, like we had this conversation already...

I just cannot imagine that some allocated resources (any) can be "not freed" in case of emergency. That would like a disaster!

Reply:
Not, just tested: neither `useEffects` cleanup nor `componentWillUnmount` will fire (just checked). 😢

Reply:
Something we need to clarify for ourselves